src/activity/activity_heartbeat_manager.rs

```diff
@@ -1,86 +1,119 @@
 use crate::errors::ActivityHeartbeatError;
 use crate::protos::coresdk::{common, ActivityHeartbeat};
 use crate::ServerGatewayApis;
-use crossbeam::channel::{bounded, unbounded, Receiver, Select, Sender, TryRecvError};
+use dashmap::mapref::one::Ref;
 use dashmap::{DashMap, DashSet};
 use futures::FutureExt;
+use std::borrow::BorrowMut;
+use std::collections::hash_map::RandomState;
 use std::convert::TryInto;
 use std::future::Future;
 use std::ops::Div;
 use std::sync::Arc;
 use std::{thread, time};
+use tokio::select;
+use tokio::sync::watch::{channel, Receiver, Sender};
+use tokio::task::JoinHandle;
 use tokio::time::{sleep, Sleep};
 
 pub(crate) struct ActivityHeartbeatManager {
     /// Core will aggregate activity heartbeats and send them to the server periodically.
-    /// This map contains latest heartbeat details for each activity that reported them.
-    heartbeat_details: DashMap<Vec<u8>, Vec<common::Payload>>,
-
-    scheduled_heartbeats: DashSet<Vec<u8>>,
-
-    heartbeat_tx: Sender<ActivityHeartbeatTask>,
-
+    /// This map contains sender channel for each activity that has an active heartbeat processor.
+    heartbeat_processors: DashMap<Vec<u8>, ActivityHeartbeatProcessorHandle>,
     shutdown_tx: Sender<bool>,
+    shutdown_rx: Receiver<bool>,
 }
 
-struct ActivityHeartbeatTask {
-    task_token: Vec<u8>,
-    next_delay: time::Duration,
-    trigger: Sleep,
+struct ActivityHeartbeatProcessorHandle {
+    heartbeat_tx: Sender<Vec<common::Payload>>,
+    join_handle: JoinHandle<()>,
+}
+
+struct ActivityHeartbeatProcessor {
+    delay: time::Duration,
```

Comment:
Slightly more specific name for `delay` might be nice - like `max_delay_between_beats` ?

Reply:
As mentioned above, let's keep it ambiguous for now as it serves dual purpose, we can choose more specific names if we decide to split up give up timeout and the delay.